Talent Competition - Employer Details + Manage Jobs

Andoliv opened the pull request.

Andoliv committed:
- .gitignore updated.

Andoliv committed:
- Task 1 implemented: Employer profile page.

Andoliv committed:
- Task 2 implemented + Optional
Display jobs as a list of cards
Bonus/Optional: Update a job, Close a job

'Copy' a Job was implemented as well.

Andoliv committed:
Revert "- Task 2 implemented + Optional"

This reverts commit 9f0a325dbe96dc8e9e499c13d49a6f171151dbaa.

Andoliv committed:
- Task 2 implemented + Optional
Display jobs as a list of cards
Bonus/Optional: Update a job, Close a job

'Copy' a Job was implemented as well.

Andoliv committed:
- Filters and Pagination fixed.

Andoliv committed:
Merge pull request #3 from Andoliv/feature/task1

Feature/task1 - Employer Details

Andoliv committed:
- Update Azure urls

Andoliv committed:
- Merge conflict

Andoliv committed:
Merge pull request #4 from Andoliv/feature/task2

Feature/task2 - Manage Jobs

Andoliv committed:
- console.log removed.

Andoliv committed:
- console.log removed.

pepemiso16 commented on App/Talent.App.WebApp/wwwroot/js/react/ReactScripts/Profile/EmployerProfile.jsx:
```diff
@@ -52,7 +52,7 @@ export default class EmployeeProfile extends React.Component {
     loadData() {
         var cookies = Cookies.get('talentAuthToken');
```
Part of this task is to utilise ES6, please be mindful when to use **let**, **const**, **var** for your variables.

kestermvp commented on App/Talent.App.WebApp/wwwroot/js/react/ReactScripts/Employer/ManageJob/ManageJob.jsx:
```diff
@@ -74,10 +111,106 @@ export default class ManageJob extends React.Component {
+                                        </span>
+                                    </div>
+                            </div>
+                            <div className="ui grid">
+                                {this.state.loadJobs.length > 0 ? 
+                                    <JobSummaryCard jobData={this.state.loadJobs}/> : 
+                                    <p>No Jobs Found</p>
+                                }
+                            </div>
+                            <div className="ui grid">
+                                <div className="center aligned sixteen wide column">
+                                    {this.state.loadJobs.length > 0 ?
+                                        <Pagination
+                                            boundaryRange={1}
+                                            defaultActivePage={this.state.activePage}
+                                            ellipsisItem={{ content: <Icon name='ellipsis horizontal' />, icon: true }}
+                                            firstItem={{ content: <Icon name='angle double left' />, icon: true }}
+                                            lastItem={{ content: <Icon name='angle double right' />, icon: true }}
+                                            prevItem={{ content: <Icon name='angle left' />, icon: true }}
+                                            nextItem={{ content: <Icon name='angle right' />, icon: true }}
+                                            siblingRange={1}
+                                            totalPages={this.state.totalPages}
+                                            onPageChange={this.handlePageChange}
+                                        /> :
+                                        null
+                                    }
+                                </div>
+                            </div>
```
Try to avoid too many conditional rendering, instead, you can create function, use it here and render what you need.

kestermvp commented on App/Talent.App.WebApp/wwwroot/js/react/ReactScripts/Employer/ManageJob/ManageJob.jsx:
```diff
@@ -74,10 +111,106 @@ export default class ManageJob extends React.Component {
+    }
+
     render() {
+        var filterOptions = [
+            { key: 'showActive', text: 'Show Active', value: 'showActive' },
+            { key: 'showClosed', text: 'Show Closed', value: 'showClosed' },
+            { key: 'showDraft', text: 'Show Draft', value: 'showDraft' },
+            { key: 'showExpired', text: 'Show Expired', value: 'showExpired' },
+            { key: 'showUnexpired', text: 'Show Unexpired', value: 'showUnexpired' }
+        ]
+
+        var sortOptions = [
+            { key: 'desc', text: 'Newest first', value: 'desc' },
+            { key: 'asc', text: 'Oldest first', value: 'asc' }
+        ]
```
Are these variables changing? If not, we can utilise **const** instead of **var**. Then if it is changing we can use **let** so you can familiarise with ES6 format. Also, don't forget to close your variable initialisation using **;.** in the end of your code

kestermvp commented on App/Talent.App.WebApp/wwwroot/js/react/ReactScripts/Employer/ManageJob/ManageJob.jsx:
```diff
@@ -74,10 +111,106 @@ export default class ManageJob extends React.Component {
+            newFilter[v] = value.includes(v);
+        });
+        
+        this.loadNewData({ filter: newFilter, activePage: 1 });
```
Can you check the following:

1. Use const or let instead of var
2. What will happen if **newFilter** doesn't contain any item? You code is fine, but you can secure it by verifying first if **newFilter** has items

kestermvp commented on App/Talent.App.WebApp/wwwroot/js/react/ReactScripts/Employer/ManageJob/ManageJob.jsx:
```diff
@@ -55,15 +58,49 @@ export default class ManageJob extends React.Component {
+                        loaderData: loaderData,
+                        totalPages: Math.ceil(res.totalCount / this.state.limit)
+                    })
+                    if (callback) callback();
```
Can you verify first the content of **res** before doing anything? You will encounter a scenario wherein, it is success but the expected response is not correct.

kestermvp commented on App/Talent.App.WebApp/wwwroot/js/react/ReactScripts/Employer/ManageJob/ManageJob.jsx:
```diff
@@ -55,15 +58,49 @@ export default class ManageJob extends React.Component {
 
     loadData(callback) {
-        var link = 'http://localhost:51689/listing/listing/getSortedEmployerJobs';
+        var link = 'https://talentservicestalentanderson.azurewebsites.net/listing/listing/getSortedEmployerJobs';
```
If this is not changing, try to use const instead of var

kestermvp commented on App/Talent.App.WebApp/wwwroot/js/react/ReactScripts/Employer/ManageJob/ManageJob.jsx:
```diff
@@ -55,15 +58,49 @@ export default class ManageJob extends React.Component {
         var cookies = Cookies.get('talentAuthToken');
-       // your ajax call and other logic goes here
+
+        $.ajax({
```
Before doing anything, can you add **try/catch** in here?

kestermvp commented on App/Talent.App.WebApp/wwwroot/js/react/ReactScripts/Employer/ManageJob/JobSummaryCard.jsx:
```diff
@@ -1,19 +1,126 @@
+    }
+
+    convertLocation(location) {
+        let locationMeta = '';
+        if (location) {
+            let { city, country } = location;
+            if (city && country) {
+                locationMeta = `${city}, ${country}`;
+            } else if (city) {
+                locationMeta = city;
+            } else if (country) {
+                locationMeta = country;
+            }
+        }
+
```
This looks good but seems too long. Can you try to implement something like this
`if (!location) return '';
    const { city, country } = location;
   locationMeta = city && country ? `${city}, ${country}` : city || country || '';`
   return locationMeta

kestermvp commented on App/Talent.App.WebApp/wwwroot/js/react/ReactScripts/Employer/ManageJob/JobSummaryCard.jsx:
```diff
@@ -1,19 +1,126 @@
+    }
+
+    init() {
+        this.props.jobData.forEach(job => {
+            let card = (
+                <Card key={job.id} fluid >
+                    <CardContent>
+                        <CardHeader>{job.title}</CardHeader>
+                        <CardMeta>{this.convertLocation(job.location)}</CardMeta>
+                        <CardDescription>{job.summary}</CardDescription>
+                    </CardContent>
+                    <CardContent extra>
+                        <div className='ui left floated'>
+                            {                                
+                                moment(job.expiryDate).isAfter(moment()) ? 
+                                    <Button content='Active' color='green' /> : 
+                                    <Button content='Expired' color='red' />
+                            }     
+                        </div>
+                        <div className='ui right floated'>
+                            <Button basic color='blue' onClick={() => this.closeJob(job.id)}>
+                                <Icon name='ban' />
+                                Close
+                            </Button>
+                            <Button basic color='blue' as={Link} to={`/EditJob/${job.id}`}>
+                                <Icon name='edit' />
+                                Edit
+                            </Button>
+                            <Button basic color='blue' as={Link} to={`/PostJob/${job.id}`}>
+                                <Icon name='copy' />
+                                Copy
+                            </Button>
+                        </div>
+                    </CardContent>
+                </Card>
+            );
+
+            this.state.cards.push(card);
+        });
```
Before directly displaying this, can you first verify if this.props.jobData has items or not? There's nothing wrong with your code, but you need to check them first because you might encounter a scenario wherein the API changes and instead of returning array the data returns an object. If that happens, your system will display an error so it is better to check and verify first.

Another thing, can you explain why are you storing the elements in a state instead of rendering them directly? Let's say you have 1000 items to handle. It wouldn;t be good if you will store all of them to a state.

kestermvp commented on App/Talent.App.WebApp/wwwroot/js/react/ReactScripts/Employer/ManageJob/JobSummaryCard.jsx:
```diff
@@ -1,19 +1,126 @@
+    }
+
+    closeJob(id) {
+        var url = 'https://talentservicestalentanderson.azurewebsites.net/listing/listing/closeJob';
+        var cookies = Cookies.get('talentAuthToken');
```
Please use const or let

kestermvp commented on App/Talent.App.WebApp/wwwroot/js/react/ReactScripts/Employer/ManageJob/JobSummaryCard.jsx:
```diff
@@ -1,19 +1,126 @@
+        var url = 'https://talentservicestalentanderson.azurewebsites.net/listing/listing/closeJob';
+        var cookies = Cookies.get('talentAuthToken');
+
+        $.ajax({
```
Can you add try/catch first and please verify if **id** has value?

kestermvp commented on App/Talent.App.WebApp/wwwroot/js/react/ReactScripts/Employer/ManageJob/JobSummaryCard.jsx:
```diff
@@ -1,19 +1,126 @@
     render() {
+        return (
+            <CardGroup itemsPerRow={2}>
+                {this.state.cards}
```
Can you call a function here instead of state?

kestermvp commented on App/Talent.App.WebApp/wwwroot/js/react/ReactScripts/Employer/CreateJob/CreateJob.jsx:
```diff
@@ -66,8 +66,8 @@ export default class CreateJob extends React.Component {
         if (param != "" || copyJobParam != "") {
-            var link = param != "" ? 'http://localhost:51689/listing/listing/GetJobByToEdit?id=' + param
-                : 'http://localhost:51689/listing/listing/GetJobForCopy?id=' + copyJobParam;
+            var link = param != "" ? 'https://talentservicestalentanderson.azurewebsites.net/listing/listing/GetJobByToEdit?id=' + param
```
Can you use const or let?

kestermvp commented on App/Talent.App.WebApp/wwwroot/js/react/ReactScripts/Employer/CreateJob/CreateJob.jsx:
```diff
@@ -95,12 +95,12 @@ export default class CreateJob extends React.Component {
     addUpdateJob() {
         var jobData = this.state.jobData;
-        console.log("data to save:", jobData);
+        // console.log("data to save:", jobData);
```
Can you add try catch here? And can you verify first if the form is valid because we can create a new job using empty form and without filling-up the required fields

Andoliv committed:
- Code review fixed.

kestermvp commented on App/Talent.App.WebApp/wwwroot/js/react/ReactScripts/Employer/ManageJob/JobSummaryCard.jsx:
```diff
@@ -72,54 +128,11 @@ export class JobSummaryCard extends React.Component {
         return (
             <CardGroup itemsPerRow={2}>
-                {this.state.cards}
+                {cards}
```
When rendering certain element or data, it's better if we can have a proper naming convention so you and other developers can be able to determine easily which are functions and which are states. For example, this **card** might be mistakenly identify as a state. You can change this into something like **renderCards** so it is easier to understad.

kestermvp commented on App/Talent.App.WebApp/wwwroot/js/react/ReactScripts/Employer/ManageJob/JobSummaryCard.jsx:
```diff
@@ -24,25 +24,81 @@ export class JobSummaryCard extends React.Component {
+            return null;
+        }
+
+        const cards = this.props.jobData.map(job => (
```
If you are handling data and going to render them, always make sure that you are treating the content correctly. Let's say for example **jobData** is an array, but, let's say the BE changes and it became an object, or accidentally return a null or empty string. If that happens, your application might display some errors. Before you map an array, always make sure that this array field is actually an array. You can check it by adding a condition like **this.props?.jobData?.map.length** and then continue with your code,

kestermvp reviewed (changes requested):
Hi @Andoliv,

Your project seems fine but there are some stuff that we need to check:

1. Can you confirm if you have any changes in the API level or in the services because I didn't see any.
2. I created a new Employer account and then login. After the successful login, my details in this form is not displaying
<img width="1178" alt="Screenshot 2024-06-16 at 3 41 27 PM" src="https://github.com/Andoliv/talent-competition/assets/168311075/a4bea9ae-9b16-4aad-9179-d53e974b13fc">
3. When I tried to click the Edit button, the form is still empty
<img width="727" alt="Screenshot 2024-06-16 at 3 41 52 PM" src="https://github.com/Andoliv/talent-competition/assets/168311075/7208b80f-678d-4498-b8ba-66f845b86ea0">

4. Inside the Post a Job page, I can be able to create a new job even if the form is empty. The UI will display the incomplete error message but it will still continue saving an empty details
<img width="489" alt="Screenshot 2024-06-16 at 3 42 39 PM" src="https://github.com/Andoliv/talent-competition/assets/168311075/c3eb5ef3-d613-4b46-8d2e-37eb29b8b75b">

5. When I removed all the default filters, the result is showing empty list of jobs.
<img width="930" alt="Screenshot 2024-06-16 at 3 45 34 PM" src="https://github.com/Andoliv/talent-competition/assets/168311075/fc230c1b-cf64-4604-9301-f2e1d0c43df4">

Andoliv commented:
> Hi @Andoliv,
> 
> Your project seems fine but there are some stuff that we need to check:
> 
> 1. Can you confirm if you have any changes in the API level or in the services because I didn't see any.
> 2. I created a new Employer account and then login. After the successful login, my details in this form is not displaying
> 
> <img alt="Screenshot 2024-06-16 at 3 41 27 PM" width="1178" src="https://private-user-images.githubusercontent.com/168311075/340063493-a4bea9ae-9b16-4aad-9179-d53e974b13fc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg2NjUxNDgsIm5iZiI6MTcxODY2NDg0OCwicGF0aCI6Ii8xNjgzMTEwNzUvMzQwMDYzNDkzLWE0YmVhOWFlLTliMTYtNGFhZC05MTc5LWQ1M2U5NzRiMTNmYy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjE3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYxN1QyMjU0MDhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hODU0MjJmMTg5ZWU4MzhiM2IyYjBkZTc1ZGE3YmVkYzZkOWI3ZDczYTdmYjcxZDZmZGY3NjRkMDQxMDFiYTQyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.SQXo7ireMYWjAHV1eRkxSCPxXz1WRGD_I29cg4kdxR8"> 3. When I tried to click the Edit button, the form is still empty <img alt="Screenshot 2024-06-16 at 3 41 52 PM" width="727" src="https://private-user-images.githubusercontent.com/168311075/340063535-7208b80f-678d-4498-b8ba-66f845b86ea0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg2NjUxNDgsIm5iZiI6MTcxODY2NDg0OCwicGF0aCI6Ii8xNjgzMTEwNzUvMzQwMDYzNTM1LTcyMDhiODBmLTY3OGQtNDQ5OC1iOGJhLTY2Zjg0NWI4NmVhMC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjE3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYxN1QyMjU0MDhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zYzY0MTMwYTdlMGI5MGI4OWVjYjZmMjAxMWUwOTk2ZTNjNDcyOTc4MDE4ZTczNTVkNmFjZWNhNDQ4NTU3N2U3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.VkCuCj5ujFfMtFx14ALl9g6e-3TwF_p-uu0kveMzoRM">
> 4. Inside the Post a Job page, I can be able to create a new job even if the form is empty. The UI will display the incomplete error message but it will still continue saving an empty details
> 
> <img alt="Screenshot 2024-06-16 at 3 42 39 PM" width="489" src="https://private-user-images.githubusercontent.com/168311075/340063577-c3eb5ef3-d613-4b46-8d2e-37eb29b8b75b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg2NjUxNDgsIm5iZiI6MTcxODY2NDg0OCwicGF0aCI6Ii8xNjgzMTEwNzUvMzQwMDYzNTc3LWMzZWI1ZWYzLWQ2MTMtNGI0Ni04ZDJlLTM3ZWIyOWI4Yjc1Yi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjE3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYxN1QyMjU0MDhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05MzFiNmJlNDdkM2Q4ZTgyM2QxM2IwMTdhZDk5N2YxN2EzNjJhOWQ2MTk2ODFhOWMzZjZjNGM0ODJhN2Y1NmQ1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.ty8TRqmXXsGTYo2lMnHy5L70CpYyauf-YzZiH7BRGjs">
> 5. When I removed all the default filters, the result is showing empty list of jobs.
> 
> <img alt="Screenshot 2024-06-16 at 3 45 34 PM" width="930" src="https://private-user-images.githubusercontent.com/168311075/340063628-fc230c1b-cf64-4604-9301-f2e1d0c43df4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg2NjUxNDgsIm5iZiI6MTcxODY2NDg0OCwicGF0aCI6Ii8xNjgzMTEwNzUvMzQwMDYzNjI4LWZjMjMwYzFiLWNmNjQtNDYwNC05MzAxLWYyZTFkMGM0M2RmNC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjE3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYxN1QyMjU0MDhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jODYwOGE2YjdiY2JjODg5ZDZjZjNkZDIzY2FiMTcwMmJkYzRmMDdjZGZkMDQ3OTBlZjgzMGNmMGY5ZmM2NTk0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.rmQixGQdhPLJEAhEM6GTSUPiH3hGSR5Bxu3UQ4WiRyo">

1. I didn't change anything in the API level before fix this issues you reported about the SignUp.
2. Fixed in the API level to put this information on Primary Contact.

Andoliv committed:
Code Review fixed:
- Backend: Primary details fixed when you sign up.
- Backend: manage jobs filters fixed.
- Frontend: blocks the creation of jobs without mandatory fields
- Frontend: this.props.JobData check if is an array
